File: app/src/main/java/jp/ddo/hotmist/unicodepad/FileChooser.java
/*
   Copyright 2018 Ryosuke839
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
       http://www.apache.org/licenses/LICENSE-2.0
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

package jp.ddo.hotmist.unicodepad;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import android.widget.Toast;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.Charset;
import java.util.Arrays;
import java.util.Enumeration;
import java.util.zip.ZipEntry;
import java.util.zip.ZipException;
import java.util.zip.ZipFile;

class FileChooser implements DialogInterface.OnClickListener, DialogInterface.OnCancelListener, ActivityCompat.OnRequestPermissionsResultCallback
{
	private final Activity activity;
	private final Listener listener;
	private String path;
	private String[] roots;
	private String[] children;

	interface Listener
	{
		void onFileChosen(String path);

		void onFileCancel();
	}

	FileChooser(Activity activity, Listener listener, String path)
	{
		this.activity = activity;
		this.listener = listener;
		this.path = path;
	}

	public boolean show()
	{
		if (Build.VERSION.SDK_INT >= 23)
		{
			if (ContextCompat.checkSelfPermission(activity, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED)
			{
				if (!ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.READ_EXTERNAL_STORAGE))
				{
					ActivityCompat.requestPermissions(activity, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, 1);
					return false;
				}
				Toast.makeText(activity, R.string.denied, Toast.LENGTH_LONG).show();
			}
		}
		onClick(null, -1);
		return true;
	}

	@Override
	public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults)
	{
		if (requestCode == 1)
			onClick(null, -1);
	}

	private ZipFile openZip(String path) throws IOException, IllegalArgumentException
	{
		ZipFile zf = new ZipFile(path);
		try
		{
			for (Enumeration<? extends ZipEntry> e = zf.entries(); e.hasMoreElements(); e.nextElement())
				;
		}
		catch (IllegalArgumentException e1)
		{
			zf.close();
			PASSED:
			{
				if (Build.VERSION.SDK_INT >= 24)
				{
					// Try to find valid charset
					for (Charset charset : Charset.availableCharsets().values())
					{
						zf = new ZipFile(path, charset);
						try
						{
							for (Enumeration<? extends ZipEntry> e = zf.entries(); e.hasMoreElements(); e.nextElement())
								;
						}
						catch (IllegalArgumentException e2)
						{
							zf.close();
							continue;
						}
						// Found
						break PASSED;
					}
				}
				// Not found
				throw new IllegalArgumentException();
			}
		}
		return zf;
	}

	@Override
	public void onClick(DialogInterface dialog, int which)
	{
		if (which != -1)
		{
			if (path.endsWith(".zip"))
			{
				if (children[which].equals(("../")))
					path = path.substring(0, path.lastIndexOf('/') + 1);
				else
				{
					//noinspection EmptyCatchBlock
					try
					{
						ZipFile zf = openZip(path);

						ZipEntry ze = zf.getEntry(children[which]);
						InputStream is = zf.getInputStream(ze);
						File of = new File(activity.getFilesDir(), String.format("%08x", ze.getCrc()) + "/" + new File(ze.getName()).getName());

						of.getParentFile().mkdirs();

						OutputStream os = new FileOutputStream(of);

						byte[] buf = new byte[256];
						int size;
						while ((size = is.read(buf)) > 0)
							os.write(buf, 0, size);

						os.close();
						is.close();
						zf.close();

						//noinspection EmptyCatchBlock
						try
						{
							if (path.startsWith(activity.getFilesDir().getCanonicalPath()))
								//noinspection ResultOfMethodCallIgnored
								new File(path).delete();
						}
						catch (IOException e)
						{
						}

						path = of.getCanonicalPath();
					}
					catch (ZipException e)
					{
						Toast.makeText(activity, R.string.cantread, Toast.LENGTH_SHORT).show();
						listener.onFileCancel();
						path = null;
						return;
					}
					catch (IllegalArgumentException e)
					{
						Toast.makeText(activity, R.string.malformed, Toast.LENGTH_SHORT).show();
						listener.onFileCancel();
						path = null;
						return;
					}
					catch (IOException e)
					{
					}
				}
			}
			else
			{
				if (path.length() != 1 && which == 0)
				{
					for (String root : roots)
						if (path.equals('/' + root))
							path = "";
				}
				if (path.length() == 0)
					path = "/";
				else
					path += children[which];
			}
		}

		if (which != -1 && path.charAt(path.length() - 1) != '/' && !path.endsWith(".zip"))
		{
			children = null;
			listener.onFileChosen(path);
			path = null;
		}
		else
		{
			//noinspection EmptyCatchBlock
			try
			{
				if (path.length() == 1)
				{
					String[] dirs = new String[4];
					dirs[0] = Environment.getExternalStorageDirectory().getCanonicalPath();
					dirs[1] = Environment.getDataDirectory().getCanonicalPath();
					dirs[2] = Environment.getDownloadCacheDirectory().getCanonicalPath();
					dirs[3] = Environment.getRootDirectory().getCanonicalPath();
					for (int i = 0; i < dirs.length; ++i)
						for (int j = 0; j < dirs.length; ++j)
							if (i != j && dirs[i].length() > 0 && dirs[j].length() > 0 && dirs[i].startsWith(dirs[j]) && new File(dirs[j]).canRead())
								dirs[i] = "";
					int cnt = 0;
					for (String dir1 : dirs)
					{
						if (!(dir1.length() > 0))
							continue;
						if (!new File(dir1).canRead())
							continue;
						++cnt;
					}
					roots = new String[cnt];
					int j = 0;
					for (String dir : dirs)
					{
						if (!(dir.length() > 0))
							continue;
						if (!new File(dir).canRead())
							continue;
						roots[j] = dir.substring(1) + '/';
						++j;
					}
					Arrays.sort(roots);
					children = roots;
				}
				else if (path.endsWith(".zip"))
				{
					try
					{
						ZipFile zf = openZip(path);
						int cnt = which != -1 ? 1 : 0;
						for (Enumeration<? extends ZipEntry> e = zf.entries(); e.hasMoreElements(); )
							if (!e.nextElement().isDirectory())
								++cnt;
						children = new String[cnt];
						int j = 0;
						if (which != -1)
							children[j++] = "../";
						for (Enumeration<? extends ZipEntry> e = zf.entries(); e.hasMoreElements(); )
						{
							ZipEntry entry = e.nextElement();
							if (entry.isDirectory())
								continue;
							children[j] = entry.getName();
							++j;
						}
						zf.close();
					}
					catch (ZipException e)
					{
						Toast.makeText(activity, R.string.cantread, Toast.LENGTH_SHORT).show();
						listener.onFileCancel();
						path = null;
						return;
					}
					catch (IllegalArgumentException e)
					{
						Toast.makeText(activity, R.string.malformed, Toast.LENGTH_SHORT).show();
						listener.onFileCancel();
						path = null;
						return;
					}
				}
				else
				{
					File f = new File(path);
					path = f.getCanonicalPath();
					if (!path.endsWith("/"))
						path += '/';
					File[] fl = f.listFiles();

					if (fl == null)
					{
						Toast.makeText(activity, R.string.cantread, Toast.LENGTH_SHORT).show();
						listener.onFileCancel();
						path = null;
						return;
					}
					int cnt = 1;
					for (File aFl1 : fl)
						if (aFl1.canRead())
							++cnt;
					children = new String[cnt];
					children[0] = "../";
					int j = 1;
					for (File aFl : fl)
					{
						if (!aFl.canRead())
							continue;
						children[j] = aFl.getName();
						if (aFl.isDirectory())
							children[j] += '/';
						++j;
					}
				}
				new AlertDialog.Builder(activity).setTitle(path.startsWith(activity.getFilesDir().getCanonicalPath()) ? path.substring(path.lastIndexOf('/') + 1) : path).setItems(children, this).setOnCancelListener(this).show();
			}
			catch (IOException e)
			{
			}
		}
	}

	@Override
	public void onCancel(DialogInterface dialog)
	{
		listener.onFileCancel();
	}
}

File: app/src/main/java/jp/ddo/hotmist/unicodepad/FileChooser.kt
/*
   Copyright 2018 Ryosuke839
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
       http://www.apache.org/licenses/LICENSE-2.0
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
package jp.ddo.hotmist.unicodepad

import android.Manifest
import android.app.Activity
import android.app.AlertDialog
import android.content.DialogInterface
import android.content.pm.PackageManager
import android.os.Build
import android.os.Environment
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.app.ActivityCompat.OnRequestPermissionsResultCallback
import androidx.core.content.ContextCompat
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.charset.Charset
import java.util.*
import java.util.zip.ZipException
import java.util.zip.ZipFile

internal class FileChooser(private val activity: Activity, private val listener: Listener, private var path: String) : DialogInterface.OnClickListener, DialogInterface.OnCancelListener, OnRequestPermissionsResultCallback {
    private var roots: Array<String> = emptyArray()
    private var children: Array<String> = emptyArray()

    internal interface Listener {
        fun onFileChosen(path: String)
        fun onFileCancel()
    }

    fun show(): Boolean {
        if (Build.VERSION.SDK_INT >= 23) {
            if (ContextCompat.checkSelfPermission(activity, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                if (!ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.READ_EXTERNAL_STORAGE)) {
                    ActivityCompat.requestPermissions(activity, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), 1)
                    return false
                }
                Toast.makeText(activity, R.string.denied, Toast.LENGTH_LONG).show()
            }
        }
        onClick(null, -1)
        return true
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == 1) onClick(null, -1)
    }

    @Throws(IOException::class, IllegalArgumentException::class)
    private fun openZip(path: String): ZipFile {
        var zf = ZipFile(path)
        try {
            val e = zf.entries()
            while (e.hasMoreElements()) {
                e.nextElement()
            }
        } catch (e1: IllegalArgumentException) {
            zf.close()
            run passed@{
                if (Build.VERSION.SDK_INT >= 24) {
                    // Try to find valid charset
                    for (charset in Charset.availableCharsets().values) {
                        zf = ZipFile(path, charset)
                        try {
                            val e = zf.entries()
                            while (e.hasMoreElements()) {
                                e.nextElement()
                            }
                        } catch (e2: IllegalArgumentException) {
                            zf.close()
                            continue
                        }
                        // Found
                        return@passed
                    }
                }
                throw IllegalArgumentException()
            }
        }
        return zf
    }

    override fun onClick(dialog: DialogInterface?, which: Int) {
        if (which != -1) {
            if (path.endsWith(".zip")) {
                if (children[which] == "../") path = path.substring(0, path.lastIndexOf('/') + 1) else {
                    try {
                        openZip(path).use { zf ->
                            val ze = zf.getEntry(children[which])
                            zf.getInputStream(ze).use { `is` ->
                                val of = File(activity.filesDir, String.format("%08x", ze.crc) + "/" + File(ze.name).name)
                                of.parentFile?.mkdirs()
                                FileOutputStream(of).use { os ->
                                    val buf = ByteArray(256)
                                    var size: Int
                                    while (`is`.read(buf).also { size = it } > 0) os.write(buf, 0, size)
                                }
                                try {
                                    if (path.startsWith(activity.filesDir.canonicalPath)) File(path).delete()
                                } catch (e: IOException) {
                                }
                                path = of.canonicalPath
                            }
                        }
                    } catch (e: ZipException) {
                        Toast.makeText(activity, R.string.cantread, Toast.LENGTH_SHORT).show()
                        listener.onFileCancel()
                        return
                    } catch (e: IllegalArgumentException) {
                        Toast.makeText(activity, R.string.malformed, Toast.LENGTH_SHORT).show()
                        listener.onFileCancel()
                        return
                    } catch (e: IOException) {
                    }
                }
            } else {
                if (path.length != 1 && which == 0) {
                    for (root in roots) if (path == "/$root") path = ""
                }
                if (path.isEmpty()) path = "/" else path += children[which]
            }
        }
        if (which != -1 && path[path.length - 1] != '/' && !path.endsWith(".zip")) {
            children = emptyArray()
            listener.onFileChosen(path)
        } else {
            try {
                if (path.length == 1) {
                    val dirs = Array(4) {""}
                    @Suppress("DEPRECATION")
                    dirs[0] = Environment.getExternalStorageDirectory().canonicalPath
                    dirs[1] = Environment.getDataDirectory().canonicalPath
                    dirs[2] = Environment.getDownloadCacheDirectory().canonicalPath
                    dirs[3] = Environment.getRootDirectory().canonicalPath
                    for (i in dirs.indices) for (j in dirs.indices) if (i != j && dirs[i].isNotEmpty() && dirs[j].isNotEmpty() && dirs[i].startsWith(dirs[j]) && File(dirs[j]).canRead()) dirs[i] = ""
                    var cnt = 0
                    for (dir1 in dirs) {
                        if (dir1.isEmpty()) continue
                        if (!File(dir1).canRead()) continue
                        ++cnt
                    }
                    roots = Array(cnt) {""}
                    var j = 0
                    for (dir in dirs) {
                        if (dir.isEmpty()) continue
                        if (!File(dir).canRead()) continue
                        roots[j] = dir.substring(1) + '/'
                        ++j
                    }
                    Arrays.sort(roots)
                    children = roots
                } else if (path.endsWith(".zip")) {
                    try {
                        openZip(path).use { zf ->
                            var cnt = if (which != -1) 1 else 0
                            run {
                                val e = zf.entries()
                                while (e.hasMoreElements()) {
                                    if (!e.nextElement().isDirectory) ++cnt
                                }
                            }
                            children = Array(cnt) { "" }
                            var j = 0
                            if (which != -1) children[j++] = "../"
                            val e = zf.entries()
                            while (e.hasMoreElements()) {
                                val entry = e.nextElement()
                                if (entry.isDirectory) {
                                    continue
                                }
                                children[j] = entry.name
                                ++j
                            }
                        }
                    } catch (e: ZipException) {
                        Toast.makeText(activity, R.string.cantread, Toast.LENGTH_SHORT).show()
                        listener.onFileCancel()
                        return
                    } catch (e: IllegalArgumentException) {
                        Toast.makeText(activity, R.string.malformed, Toast.LENGTH_SHORT).show()
                        listener.onFileCancel()
                        return
                    }
                } else {
                    val f = File(path)
                    f.canonicalPath.let {
                        path = it + if (!it.endsWith("/")) "/" else ""
                    }
                    val fl = f.listFiles()
                    if (fl == null) {
                        Toast.makeText(activity, R.string.cantread, Toast.LENGTH_SHORT).show()
                        listener.onFileCancel()
                        return
                    }
                    var cnt = 1
                    for (aFl1 in fl) if (aFl1.canRead()) ++cnt
                    children = Array(cnt) {""}
                    children[0] = "../"
                    var j = 1
                    for (aFl in fl) {
                        if (!aFl.canRead()) continue
                        children[j] = aFl.name + if (aFl.isDirectory) "/" else ""
                        ++j
                    }
                }
                AlertDialog.Builder(activity).setTitle(if (path.startsWith(activity.filesDir.canonicalPath)) path.substring(path.lastIndexOf('/') + 1) else path).setItems(children, this).setOnCancelListener(this).show()
            } catch (e: IOException) {
            }
        }
    }

    override fun onCancel(dialog: DialogInterface) {
        listener.onFileCancel()
    }
}